torchdata/datapipes/iter/transform/callable.py

```diff
@@ -205,3 +206,162 @@ def __len__(self) -> int:
         if isinstance(self.datapipe, Sized):
             return len(self.datapipe)
         raise TypeError(f"{type(self).__name__} instance doesn't have valid length")
+
+
+@functional_datapipe("islice")
+class ISliceIterDataPipe(IterDataPipe[T_co]):
+    r"""
+    returns a slice of elements in input DataPipe via start/stop/step or indices (functional name: ``islice``).
+
+    Args:
+        datapipe: IterDataPipe with iterable elements
+        index: a single start index for the slice or a list of indices to be returned instead of a start/stop slice
+        stop: the slice stop. ignored if index is a list
+        step: step to be taken from start to stop. ignored if index is a list
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> dp = IterableWrapper([(0, 10, 100), (1, 11, 111), (2, 12, 122), (3, 13, 133), (4, 14, 144)])
+        >>> islice_dp = dp.islice(0, 2)
+        >>> list(islice_dp)
+        [(0, 10), (1, 11), (2, 12), (3, 13), (4, 14)]
+    """
+    datapipe: IterDataPipe
+
+    def __init__(
+        self,
+        datapipe: IterDataPipe,
+        index: Union[int, List[Hashable]],
+        stop: Optional[int] = None,
+        step: Optional[int] = None,
+    ) -> None:
+        super().__init__()
+        self.datapipe = datapipe
+
+        self.index = index
+        self.stop = stop
+        self.step = step
```

Comment:
Could we add a check here to validate list of indices should be mutually exclusive from `stop` and `step`?